### DISCUSSION TITLE
How to return translations from functions outside of React components

### DISCUSSION THREAD

**Discussion by kuchejak** [amannn/next-intl#1575] [Q&A] [answered]
Hi, I have read the [How (not) to use translations outside of React components](https://next-intl-docs.vercel.app/blog/translations-outside-of-react-components) blog post and I guess I understand the intentions, but I can't figure out how to properly use translation in some server functions, that are neither server actions or React component (but might be called from both).

For example I might have a function `checkSomething` that checks that the user can perform certain operation and if not, returns an error describing what went wrong. Now I need that error to be translated for the user.

Simple way would be to return `keyof Messages["NameOfTheMessage"]` and then translate it in the server action or React component that calls the `checkSomething` function. 

But what type should I return when I also want to pass in parametr for the translate method? 
```
        t("problematicMember", {
          name: `${member.firstname} ${member.surname}`,
        }),
```

The typescript page [typescript integration page](https://next-intl-docs.vercel.app/docs/workflows/typescript) did not help.

And is this approach need, when the function is called only from a server action or React server component? I'm not sure if the problems described in the blog post apply there (but I'm quite new to Next.js so I'm not sure about anything). 

Thanks!

**Comment by amannn** [comment, marked as answer]
This seems similar to how [Zod can be used in Server Actions, while using translations](https://next-intl-docs.vercel.app/docs/environments/actions-metadata-route-handlers#server-actions-zod).

What happens is basically:
1. Zod returns a data structure that describes an error
2. The calling Server Action retrieves `t` from `await getTranslations` and turns the data structure into a human-readable error message (could also be a component in your case)

It's essentially the same pattern that's also described in the blog post:

> we can format messages during the rendering phase of React, turning data structures into human readable strings

So I'd try to avoid coupling your `checkSomething` function to translations, but it could for example return error codes (e.g. `code: 'PROBLEMATIC_MEMBER'`), potentially with metadata (`data: {member: …}`).

Does that help?

**Reply by kuchejak** [reply]
Yes, that makes sense, thank you!
